execution-engine/src/wasi/strace.rs

```diff
@@ -9,30 +9,99 @@
 //! See the `LICENSE_MIT.markdown` file in the Veracruz root directory for
 //! information on licensing and copyright.
 
-#![allow(unused_variables)] // while most of the functions are unimplemented
-
 use super::common::MemoryHandler;
 use crate::fs::FileSystemResult;
 use std::fmt;
 
-pub struct Strace {}
+enum TraceState {
+    Func,
+    Args,
+    Result,
+}
+
+pub struct Strace {
+    enabled: bool,
+    state: TraceState,
+}
+
+// Convert a vector of bytes into a printable ASCII string.
+// The string may contain readable text, which is useful to display,
+// but it may also contain binary data, so we cannot decode it as UTF-8.
+fn strace_string(bytes: &[u8], max: usize) -> String {
+    let mut res = String::from("\"");
+    let n = if bytes.len() > max { max } else { bytes.len() };
+    for i in 0..n {
```

Comment:
I wonder whether this can be refactored to use the `windows` method on slices:

```rust
for &[current, next] in bytes.iter().windows(2) {
  if 0x20 <= current && current <= 0x7f {
    ...
  }
  ...
  } else if b'0' <= next && next <= b'9' {
    // The following character is a digit, so use three octal digits.
    res.push_str(&format!("\\{:03o}:, current);
  }
  ...
}
```

...or something similar.

Reply:
But how would I handle the last character of the slice?